lib/devShell.nix

```diff
@@ -0,0 +1,38 @@
+{ lib
+, mkShell
+, rustPlatform
+, cargo
+, rustc
+, pkgs
+}:
+
+{ checks ? {}
+, inputsFrom ? []
+, packages ? []
+, rust-analyzer ? pkgs.rust-analyzer
+, ...
+}@args:
+let
+  inherit (builtins) removeAttrs;
+
+  cleanedArgs = removeAttrs args [
+    "checks"
+    "inputsFrom"
+    "nativeBuildInputs"
+    "rust-analyzer"
+  ];
+in
+mkShell (cleanedArgs // {
+  inputsFrom = builtins.attrValues checks ++ inputsFrom;
+
+  packages =
+    [
+      rustc
+      cargo
+    ]
+    ++ lib.optional (rust-analyzer != null) rust-analyzer
+    ++ packages;
+}
+// lib.optionalAttrs (rust-analyzer != null) {
+  RUST_SRC_PATH = rustPlatform.rustLibSrc;
```

Comment:
Unfortunately we can't use `rustPlatform` here since it will not know about any toolchains overlayed during instantiation (e.g. `craneLib.overrideToolchain`). 

I'm not sure of the best way to get this. I believe that any `rustc` derivations (e.g. via `oxalica/rust-overlay`) which contain the `src` component _should_ make it available somewhere, but that starts making assumptions about how toolchains are being installed/configured.

If there's another decent heuristic we can use I'd be happy to use it, but if we can't find one, I'd rather we leave this part up to whomever needs it (though we can also add it back in if we find such a heuristic later)

Reply:
Hmm, I see. This variable is necessary to get `rust-analyzer` working, let me think about this.

Reply:
So let me check if I understand this correctly:

The issue is that `pkgs.rustPlatform` may not be the correct `rustPlatform` given `craneLib.overrideToolchain` can be used to select a different Rust platform, and the different Rust platform may not include a `rustPlatform` or `rustLibSrc` attribute?

Reply:
For the `rustWithWasiTarget` toolchain/override here, the path `${rustWithWasiTarget.availableComponents.rust-src}/lib/rustlib/src/rust/library/` is equivalent to `rustPlatform.rustLibSrc`:

https://github.com/ipetkov/crane/blob/6aaaa996c4e039b85dafc1ed9d0f4f075c49b50a/examples/custom-toolchain/flake.nix#L31-L39

Maybe we could check if there's a `rustc.availableComponents` attribute and use the `oxalica/rust-overlay` source in that case, or `rustPlatform.rustLibSrc` otherwise? I'm not sure it's bulletproof though, what do you think?

Reply:
Alternatively, we could make `rustLibSrc` an explicit argument to `devShell` and only set `RUST_SRC_PATH` if it's given.

Reply:
> The issue is that `pkgs.rustPlatform` may not be the correct `rustPlatform` given `craneLib.overrideToolchain` can be used to select a different Rust platform, and the different Rust platform may not include a `rustPlatform` or `rustLibSrc` attribute?

Yep this is correct. The only time `pkgs.rustPlatform` would change is if someone overlayed a different `rustc` or `cargo` version into `nixpkgs`, but `crane`'s view of `rustc` and `cargo` can be influenced independently (via `.overrideToolchain`).

We have no guarantees about what the `rustc`/`cargo` values passed into `.overrideToolchain` are (besides the presumption that they contain their own respective binary which behaves as one might generally expect)

> For the rustWithWasiTarget toolchain/override here, the path ${rustWithWasiTarget.availableComponents.rust-src}/lib/rustlib/src/rust/library/ is equivalent to rustPlatform.rustLibSrc:

They are equivalent, yes, but not identical, meaning they can come from two entirely different toolchains (notably nixpkgs usually lags around one stable version behind Rust releases)

---

Given these challenges, for the time being (this PR) I would prefer we _neither include `rust-analyer` nor set `RUSTC_SRC_PATH`_: if we cannot get things exactly right (e.g. point `rust-analyer` to the "correct" source for the toolchain) we should not make an incorrect guess and give someone a debugging headache.

Totally happy to revisit this in the future; if the API would need to change to support it better we can also explore that in a future PR since I think there's a lot of value what's in here already!

Reply:
Done!